pkg/egresspolicy/manager.go

```diff
@@ -0,0 +1,461 @@
+//  Copyright 2021 Authors of Cilium
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package egresspolicy
+
+import (
+	"fmt"
+	"net"
+
+	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	k8sUtils "github.com/cilium/cilium/pkg/k8s/utils"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/maps/egressmap"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "egresspolicy")
+)
+
+// The egresspolicy manager stores the internal data tracking the policy
+// and pod mappings. And also hooks up all the callbacks to update the egress
+// map accordingly.
+type Manager struct {
+	mutex lock.Mutex
+
+	// Stores pod to policy mapping
+	policyPods map[podID][]policyID
+	// Stores policy configs indexed by policyID
+	policyConfigs map[policyID]*EgressPolicyConfig
+	// Stores podId to pod metadata mapping
+	podDataStore map[podID]*podMetadata
+
+	storeGetter StoreGetter
+}
+
+func NewEgressPolicyManager() *Manager {
+	return &Manager{
+		policyPods:    make(map[podID][]policyID),
+		policyConfigs: make(map[policyID]*EgressPolicyConfig),
+		podDataStore:  make(map[podID]*podMetadata),
+	}
+}
+
+type StoreGetter interface {
+	GetStore(name string) cache.Store
+}
+
+func (manager *Manager) RegisterGetStores(sg StoreGetter) {
+	manager.storeGetter = sg
+}
+
+// Event handlers
+
+// AddEgressPolicy parses the given policy config, and updates internal state with the config fields.
+// returns bool indicates if policy is added, err inidates first encountered error
+func (manager *Manager) AddEgressPolicy(config EgressPolicyConfig) (bool, error) {
+	manager.mutex.Lock()
+	defer manager.mutex.Unlock()
+
+	_, ok := manager.policyConfigs[config.id]
+	if ok {
+		log.WithField(
+			logfields.CiliumEgressNATPolicyName, config.id.Name).Warn("CiliumEgressNATPolicy already exists and is not re-added")
+		return false, fmt.Errorf("%v already exists", config.id.Name)
```

Comment:
I don't think we need to encode the name of the policy in the error msg. The reason is because it's desirable to have fixed error msgs, instead of dynamic. In this case, we already know which policy already exists because of the log msg immediately above the return. 

```suggestion
		return false, errors.New("already exists",)
```

or

If we need to act on a specific error type, then it might also make sense to do:

```suggestion
              var errExists = errors.New("already exists") // Somewhere defined at a global scope
		return false, errExists
```

Reply:
Fixed.